### PR TITLE
trainer.test()

### LINKED ISSUE #10363: trainer.test

### PR DESCRIPTION
**Pull request by JiayiFeng**
Fixes #10363

### CODE REVIEW

**Comment by JiayiFeng** [comment]
Two problems I have encountered:

1. The `train_program` may have multiple return values, we need a method to point out which one is the loss.

https://github.com/JiayiFeng/Paddle/blob/21be154df583ae8ba032379878018f90173da563/python/paddle/fluid/trainer.py#L77-L88

2. In the test, we get metric results of every batch and accumulate them. However, if users return some variables which are not metrics and cannot be accumulated, what shall we do?

File: python/paddle/fluid/trainer.py [hunk 1]
```diff
@@ -225,26 +239,59 @@ def _train_by_executor(self, num_epochs, event_handler, reader, feed_order):
...
+            accumulated_loss = 0
+            count = 0
+            for data in reader():
+                loss, = exe.run(program=self.test_program,
+                                feed=feeder.feed(data),
+                                fetch_list=fetch_list)
+                accumulated_loss += loss[0]
+                count += 1
+
+            return accumulated_loss / count
```
Comment (jetfuel) [inline python/paddle/fluid/trainer.py]: Should we return Accuracy as well?

**Comment by daming-lu** [comment]
Based on our discussion, we define the 1st element in the returned tuple as `loss`. The rest are evaluation metrics. So a sample returned result would look like
`python
(loss, accuracy, f1_score, ...)
`

File: python/paddle/fluid/trainer.py [hunk 2]
```diff
@@ -168,8 +172,17 @@ def train(self,
...
+        """
+        Test the model on given test data
+
+        Args:
+            reader: The reader that yields test data.
+            feed_order: Feeding order of reader. None will following the defining
+                order in program
+        """
+
+        return self._test_by_executor(reader, feed_order, self.test_outputs)
```
Comment (jacquesqiao) [inline python/paddle/fluid/trainer.py]: Why put all code into a separate function?

Comment (JiayiFeng) [reply python/paddle/fluid/trainer.py]: Maybe in the future, we need to support `_test_by_parralle_executor`. Here is a reserved interface for switching between multiple executors.

File: python/paddle/fluid/tests/book/high-level-api/word2vec/no_test_word2vec_new_api.py [hunk 3]
```diff
@@ -80,8 +80,11 @@ def inference_program(is_sparse):
 
 
 def train_program(is_sparse):
-    next_word = fluid.layers.data(name='nextw', shape=[1], dtype='int64')
+    # The declaration of 'next_word' must be after the invoking of inference_program,
+    # or the data input order of train program would be [next_word, firstw, secondw,
+    # thirdw, forthw], which is not correct.
     predict_word = inference_program(is_sparse)
```
Comment (jetfuel) [inline python/paddle/fluid/tests/book/high-level-api/word2vec/no_test_word2vec_new_api.py]: So is it accurate to say that "the very line of train_program should be getting the infererence_program"?

File: python/paddle/fluid/trainer.py [hunk 4]
```diff
@@ -225,26 +238,59 @@ def _train_by_executor(self, num_epochs, event_handler, reader, feed_order):
...
                 event_handler(BeginEpochEvent(epoch_id))
                 for step_id, data in enumerate(reader()):
                     event_handler(BeginStepEvent(epoch_id, step_id))
                     exe.run(feed=feeder.feed(data), fetch_list=[])
                     event_handler(EndStepEvent(epoch_id, step_id))
                 event_handler(EndEpochEvent(epoch_id))
+
+    def _test_by_executor(self, reader, feed_order, fetch_list):
+        with executor.scope_guard(self.scope):
+            feed_var_list = build_feed_var_list(self.test_program, feed_order)
```
Comment (jetfuel) [inline python/paddle/fluid/trainer.py]: Trainer. train takes in `feed_order`, Will it be possible to return the `feed_order` so we don't need to calculate it?

Comment (jetfuel) [inline python/paddle/fluid/trainer.py]: Trainer. train takes in `feed_order`, Will it be possible to re-use the feed_order so we don't need to calculate it?
